lib/liquid/static_registers.rb

```diff
@@ -1,36 +1,35 @@
 # frozen_string_literal: true
 
+require 'forwardable'
+
 module Liquid
   class StaticRegisters
+    extend Forwardable
+
     attr_reader :static, :registers
 
+    def_delegators :@cache, :[], :key?
+
     def initialize(registers = {})
       @static    = registers.is_a?(StaticRegisters) ? registers.static : registers
       @registers = {}
+
+      @cache = @static.dup
     end
 
     def []=(key, value)
       @registers[key] = value
-    end
-
-    def [](key)
-      if @registers.key?(key)
-        @registers[key]
-      else
-        @static[key]
-      end
+      @cache[key] = value
     end
 
     def delete(key)
-      @registers.delete(key)
+      @registers.delete(key).tap do
+        @static.dup.merge(@registers)
```

Comment:
I don't understand this tap block, since `tap` doesn't return the result of the block so it looks like this block is just building an object that won't get used.

Reply:
I had this before that I can put back if you desire?
```rb
def delete(key)
  value = @registers.delete(key)
  @static.dup.merge(@registers)
  value
end
```

Reply:
Actually I'm maintaining a behaviour that is tested, but we are doing a breaking change in a subsequent PR. 

I wouldn't mind if we preferred to do further breaking changes.

Reply:
A `@static.dup.merge(@registers)` line is the same thing, create a new hash but then doesn't use it.

Reply:
Haaa it's because I'm dumb and forgot to assign it to `@cache`